### PR TITLE
[RemoveDIs] Update DIBuilder C API with DbgRecord functions.

### PR DESCRIPTION
**Pull request by CarlosAlbertoEnciso**
The DIBuilder C API was changed to deal with DbgRecord functions:

https://github.com/llvm/llvm-project/pull/84915
https://github.com/llvm/llvm-project/pull/85657
https://github.com/llvm/llvm-project/pull/92417#issuecomment-2120078326

As discussed by @nikic and @OCHyams:

https://github.com/llvm/llvm-project/pull/92417#issuecomment-2144505440

> For the intrinsic-inserting functions, do you think we should:
>
> a) mark as deprecated but otherwise leave them alone for now.
> b) mark as deprecated and change their behaviour so that they
>    do nothing and return nullptr.
> c) outright delete them (it sounds like you're voting this one,
>    but just wanted to double check).

This patch implements option (c).

### CODE REVIEW

File: llvm/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -220,7 +220,43 @@ Changes to the C API
...
+
+  * ``LLVMDIBuilderInsertDeclareBefore``
+  * ``LLVMDIBuilderInsertDeclareAtEnd``
+  * ``LLVMDIBuilderInsertDbgValueBefore``
+  * ``LLVMDIBuilderInsertDbgValueAtEnd``
+
+  * ``LLVMDIBuilderInsertDeclareIntrinsicBefore``
+  * ``LLVMDIBuilderInsertDeclareIntrinsicAtEnd``
+  * ``LLVMDIBuilderInsertDbgValueIntrinsicBefore``
+  * ``LLVMDIBuilderInsertDbgValueIntrinsicAtEnd``
```
Comment (OCHyams) [inline llvm/docs/ReleaseNotes.rst:259]: You don't need to mention these as these were ones I added that we're removing, they are not part of the original C API.

File: llvm/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -220,7 +220,43 @@ Changes to the C API
...
+
+  * ``LLVMPositionBuilderBeforeDbgRecords``
+  * ``LLVMPositionBuilderBeforeInstrAndDbgRecords``
+
+  Same as ``LLVMPositionBuilder`` and ``LLVMPositionBuilderBefore`` except the
+  insertion position is set to before the debug records that precede the target
+  instruction. ``LLVMPositionBuilder`` and ``LLVMPositionBuilderBefore`` are
+  unchanged.
+
+  See the `debug info migration guide <https://llvm.org/docs/RemoveDIsDebugInfo.html>`_ for more info.
```
Comment (OCHyams) [inline llvm/docs/ReleaseNotes.rst:234]: Is this change needed?

File: llvm/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -220,7 +220,43 @@ Changes to the C API
...
+
+* Added the following functions to get/set the new non-instruction debug info format.
+
+  * ``LLVMIsNewDbgInfoFormat``
+  * ``LLVMSetIsNewDbgInfoFormat``
+
+* Added the following functions (no plans to deprecate) to insert a debug record
```
Comment (OCHyams) [inline llvm/docs/ReleaseNotes.rst:241]: "no plans to deprecate" would be the default assumption so I don't think you need to mention it here.

File: llvm/docs/RemoveDIsDebugInfo.md [hunk 4]
```diff
@@ -140,34 +140,36 @@ Any tests downstream of the main LLVM repo that test the IR output of LLVM may b
...
+-----------------
+LLVMDIBuilderInsertDeclareBefore   # Insert a debug record (new debug info format) instead of a debug intrinsic (old debug info format).
+LLVMDIBuilderInsertDeclareAtEnd    # Same as above.
+LLVMDIBuilderInsertDbgValueBefore  # Same as above.
+LLVMDIBuilderInsertDbgValueAtEnd   # Same as above.
 
 LLVMDIBuilderInsertDeclareIntrinsicBefore   # Insert a debug intrinsic (old debug info format).
 LLVMDIBuilderInsertDeclareIntrinsicAtEnd    # Same as above.
 LLVMDIBuilderInsertDbgValueIntrinsicBefore  # Same as above.
 LLVMDIBuilderInsertDbgValueIntrinsicAtEnd   # Same as above.
```
Comment (OCHyams) [inline llvm/docs/RemoveDIsDebugInfo.md:153]: Same as before - these ones don't need to be mentioned at all.

File: llvm/docs/ReleaseNotes.rst [hunk 5]
```diff
@@ -220,7 +220,43 @@ Changes to the C API
...
+  * ``LLVMPositionBuilderBeforeInstrAndDbgRecords``
+
+  Same as ``LLVMPositionBuilder`` and ``LLVMPositionBuilderBefore`` except the
+  insertion position is set to before the debug records that precede the target
+  instruction. ``LLVMPositionBuilder`` and ``LLVMPositionBuilderBefore`` are
+  unchanged.
+
+  See the `debug info migration guide <https://llvm.org/docs/RemoveDIsDebugInfo.html>`_ for more info.
+
+* Added the following functions to get/set the new non-instruction debug info format.
```
Comment (OCHyams) [inline llvm/docs/ReleaseNotes.rst:236]: IMO worth adding a note that these will be deprecated in the future and they're just a transition aid.

File: llvm/tools/llvm-c-test/debuginfo.c [hunk 6]
```diff
@@ -137,38 +140,13 @@ int llvm_test_dibuilder(bool NewDebugInfoFormat) {
...
                                          42, Int64Ty, true, 0);
 
-  if (LLVMIsNewDbgInfoFormat(M))
-    LLVMDIBuilderInsertDeclareAtEnd(
-        DIB, LLVMConstInt(LLVMInt64Type(), 0, false), FooParamVar1,
-        FooParamExpression, FooParamLocation, FooEntryBlock);
-  else
-    LLVMDIBuilderInsertDeclareIntrinsicAtEnd(
-        DIB, LLVMConstInt(LLVMInt64Type(), 0, false), FooParamVar1,
-        FooParamExpression, FooParamLocation, FooEntryBlock);
```
Comment (OCHyams) [inline llvm/tools/llvm-c-test/debuginfo.c:147]: The test should now call the `LLVMDIBuilderInsertDeclareRecord...` style functions. e.g. at this site:

delete the `if ... else ...`, and in its place instead just call `LLVMDIBuilderInsertDeclareRecordAtEnd(LLVMInt64Type(), 0, false), FooParamVar1,
        FooParamExpression, FooParamLocation, FooEntryBlock);`

It looks like you've done that for the final block in the test but not the first 3.

Otherwise we're losing test coverage.

File: llvm/include/llvm-c/DebugInfo.h [hunk 7]
```diff
@@ -1261,40 +1261,6 @@ LLVMMetadataRef LLVMDIBuilderCreateTempGlobalVariableFwdDecl(
...
- * \param VarInfo     The variable's debug info descriptor.
- * \param Expr        A complex location expression for the variable.
- * \param DebugLoc    Debug info location.
- * \param Instr       Instruction acting as a location for the new intrinsic.
- */
-LLVMValueRef LLVMDIBuilderInsertDeclareIntrinsicBefore(
-    LLVMDIBuilderRef Builder, LLVMValueRef Storage, LLVMMetadataRef VarInfo,
-    LLVMMetadataRef Expr, LLVMMetadataRef DebugLoc, LLVMValueRef Instr);
 /**
  * Soon to be deprecated.
```
Comment (OCHyams) [inline llvm/include/llvm-c/DebugInfo.h:1265]: Please can you remove the `Soon to be deprecated` notes from all the `Record` versions? The plan is now to keep the `Record` versions so these comments are out of date and misleading.

While you're there please could you add another note to the functions indicating that the debug format can be switched later after inserting the records using LLVMSetIsNewDbgInfoFormat, if needed for legacy or transitionary reasons.

File: llvm/bindings/ocaml/debuginfo/debuginfo_ocaml.c [hunk 8]
```diff
@@ -969,45 +969,6 @@ value llvm_dibuild_create_parameter_variable_bytecode(value *argv, int arg) {
   );
 }
 
-value llvm_dibuild_insert_declare_before_native(value Builder, value Storage,
-                                                value VarInfo, value Expr,
-                                                value DebugLoc, value Instr) {
-  LLVMDbgRecordRef Value = LLVMDIBuilderInsertDeclareBefore(
```
Comment (OCHyams) [inline llvm/bindings/ocaml/debuginfo/debuginfo_ocaml.c:975]: I think we want to keep these functions and change `LLVMDIBuilderInsertDeclareBefore` for the record versions (same below).

File: llvm/bindings/ocaml/debuginfo/llvm_debuginfo.ml [hunk 9]
```diff
@@ -592,26 +592,6 @@ external dibuild_create_parameter_variable :
...
-external dibuild_insert_declare_at_end :
-  lldibuilder ->
-  storage:Llvm.llvalue ->
-  var_info:Llvm.llmetadata ->
-  expr:Llvm.llmetadata ->
-  location:Llvm.llmetadata ->
-  block:Llvm.llbasicblock ->
-  Llvm.lldbgrecord
-  = "llvm_dibuild_insert_declare_at_end_bytecode" "llvm_dibuild_insert_declare_at_end_native"
-
```
Comment (OCHyams) [inline llvm/bindings/ocaml/debuginfo/llvm_debuginfo.ml:614]: See comments above from debuginfo_ocml.c - I don't these should be deleted

File: llvm/docs/ReleaseNotes.rst (hunk 2 above)
Comment (CarlosAlbertoEnciso) [reply llvm/docs/ReleaseNotes.rst:234]: Just to add clarity to the change.

File: llvm/docs/ReleaseNotes.rst (hunk 3 above)
Comment (CarlosAlbertoEnciso) [reply llvm/docs/ReleaseNotes.rst:241]: Removed.

File: llvm/docs/ReleaseNotes.rst (hunk 5 above)
Comment (CarlosAlbertoEnciso) [reply llvm/docs/ReleaseNotes.rst:236]: Added a note.

**Comment by pogo59** [comment]
Is this removing any C APIs that existed in the previous release? That's a no-no.

File: llvm/include/llvm-c/DebugInfo.h (hunk 7 above)
Comment (CarlosAlbertoEnciso) [reply llvm/include/llvm-c/DebugInfo.h:1265]: Added the following note:
```
 *
 * The debug format can be switched later after inserting the records using
 * LLVMSetIsNewDbgInfoFormat, if needed for legacy or transitionary reasons.
 *
```

**Comment by OCHyams** [comment]
> Is this removing any C APIs that existed in the previous release? That's a no-no.

Yes and no. It's removing some functions that I added during this release (unambiguously OK) but it's also removing some functions that insert debug intrinsics (as opposed to debug records) in accordance with the discussion linked by Carlos in the summary (https://github.com/llvm/llvm-project/pull/92417#issuecomment-2120078326 is the important one).

It was discussed whether the existing functions should be deprecated first, but @nikic voted in favour of deleting them.

File: llvm/bindings/ocaml/debuginfo/debuginfo_ocaml.c (hunk 8 above)
Comment (CarlosAlbertoEnciso) [reply llvm/bindings/ocaml/debuginfo/debuginfo_ocaml.c:975]: Restoring the deleted functions and changed them to use the record versions.

File: llvm/bindings/ocaml/debuginfo/llvm_debuginfo.ml (hunk 9 above)
Comment (CarlosAlbertoEnciso) [reply llvm/bindings/ocaml/debuginfo/llvm_debuginfo.ml:614]: Restoring the deleted functions.

**Comment by CarlosAlbertoEnciso** [comment]
@OCHyams Thanks for your valuable comments. Uploaded new patch to address them.

**Comment by CarlosAlbertoEnciso** [comment]
@alan-j-hu: looking at the original rework of the C-API, it seems that you run the unit test for the OCaml bindings.
https://github.com/llvm/llvm-project/pull/92417#issuecomment-2116328768

I followed https://ocaml.org/docs/installing-ocaml to install OCaml and added `LLVM_ENABLE_BINDINGS` to enable the OCaml buildings.

Running `ninja check-llvm-bindings` I get the following error:

File "....../test/Bindings/OCaml/Output/debuginfo.ml.tmp/debuginfo.ml", line 19, characters 20-50:
19 | let null_metadata = Llvm_debuginfo.llmetadata_null ()
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound module "Llvm_debuginfo"

Is there any more specific instructions that I can follow for the OCaml bindings. Thanks

File: llvm/docs/ReleaseNotes.rst [hunk 10]
```diff
@@ -220,7 +220,38 @@ Changes to the C API
...
+  * ``LLVMSetIsNewDbgInfoFormat``
+
+* Added the following functions to insert a debug record (new debug info format).
+
+  * ``LLVMDIBuilderInsertDeclareRecordBefore``
+  * ``LLVMDIBuilderInsertDeclareRecordAtEnd``
+  * ``LLVMDIBuilderInsertDbgValueRecordBefore``
+  * ``LLVMDIBuilderInsertDbgValueRecordAtEnd``
+
+* Deleted the following functions that inserted a debug intrinsic (old debug info format)
```
Comment (OCHyams) [inline llvm/docs/ReleaseNotes.rst:249]: nit: missing full stop.

It's a bit repetitive but we probably want `See the `debug info migration guide <https://llvm.org/docs/RemoveDIsDebugInfo.html>`_ for more info.` included in each one of these descriptive bullet points. What do you think?

**Review by OCHyams — COMMENTED** [review]
Just one nit and otherwise LGTM, thanks for this @CarlosAlbertoEnciso 

We'll want to wait for @pogo59 to give the thumbs-up and @nikic to take a quick look (though I'm pretty confident that these changes line up with what @nikic outlines in the other threads).

**Review by nikic — APPROVED** [review]
LGTM

File: llvm/docs/ReleaseNotes.rst (hunk 10 above)
Comment (CarlosAlbertoEnciso) [reply llvm/docs/ReleaseNotes.rst:249]: What I would suggest is something like this:

Changes to the C API
--------------------
...
**Note:** The following changes are due to the removal of the debug info
intrinsics from LLVM and to the introduction of debug records into LLVM.
They are described in detail in the `debug info migration guide <https://llvm.org/docs/RemoveDIsDebugInfo.html>`_.

* Added the following functions to insert before the indicated instruction but
....

Comment (OCHyams) [reply llvm/docs/ReleaseNotes.rst:249]: That moves the info further away from the place someone would be looking, and requires these changes always sit at the bottom of the list. These are very minor concerns though so I'm happy to go with whatever you prefer here (don't forget the full stop).

**Comment by alan-j-hu** [comment]
@CarlosAlbertoEnciso IIRC the build files are missing the OCaml debuginfo module in some list, so you have to explicitly build the target `ocaml_llvm_debuginfo`, which is not automatically built when you run `check-llvm-bindings-ocaml `.

**Comment by pogo59** [comment]
OK I've looked up the C API evolution rules and removing APIs is okay given a release note, which you have. So, no objection from me.

I do have to wonder whether `LLVM[Set]IsNewDbgInfoFormat` is useful? Given that the APIs for inserting old info no longer exist.

File: llvm/docs/ReleaseNotes.rst (hunk 10 above)
Comment (CarlosAlbertoEnciso) [reply llvm/docs/ReleaseNotes.rst]: @pogo59, @nikic any opinion on this? Thanks.

File: llvm/include/llvm-c/DebugInfo.h [hunk 11]
```diff
@@ -1420,46 +1321,11 @@ LLVMDbgRecordRef LLVMDIBuilderInsertDbgValueRecordBefore(
...
-    LLVMDIBuilderRef Builder, LLVMValueRef Val, LLVMMetadataRef VarInfo,
-    LLVMMetadataRef Expr, LLVMMetadataRef DebugLoc, LLVMBasicBlockRef Block);
-/**
- * Soon to be deprecated.
- * Only use in "new debug mode" (LLVMIsNewDbgInfoFormat() is true).
- * See https://llvm.org/docs/RemoveDIsDebugInfo.html#c-api-changes
+ * The debug format can be switched later after inserting the records using
+ * LLVMSetIsNewDbgInfoFormat, if needed for legacy or transitionary reasons.
  *
  * Insert a new llvm.dbg.value intrinsic call at the end of the given basic
```
[the remark this answers is not in the document]
Comment (CarlosAlbertoEnciso) [reply llvm/include/llvm-c/DebugInfo.h]: @nikic Thanks for your valuable review. Replaced the 'intrinsic' occurrences.

**Comment by OCHyams** [comment]
> I do have to wonder whether LLVM[Set]IsNewDbgInfoFormat is useful? Given that the APIs for inserting old info no longer exist.

We needed it in tests while transitioning. There's a chance it might be useful still for downsteam projects still atm (allows a producer flip back to the old format after inserting the records). It won't be useful in the future once intrinsics have been completely removed (hence the planned deprecation).

I'm not sure it's doing much harm and could potentially be helpful, but I don't have strong feelings about it. If you think that the clutter isn't warranted we could remove that too (maybe in a separate commit to keep things moving?).

**Comment by pogo59** [comment]
> I'm not sure it's doing much harm and could potentially be helpful, 

Just wanted to make sure there was some potential purpose. If you plan to remove it later that's fine.

File: llvm/docs/ReleaseNotes.rst (hunk 10 above)
Comment (pogo59) [reply llvm/docs/ReleaseNotes.rst]: These notes are for a group of related changes. The "Note" seems okay to me. Another tactic would be to put this group of changes into a new subsection, which could help other release-note writers realize what's going on. That said, C API changes aren't super common and I'm equally fine with just having the Note.

**Review by alan-j-hu — APPROVED** [review]
The OCaml tests pass, so I approve the OCaml-related changes. (Of course, I'm not an expert on the C API part.)

**Comment by CarlosAlbertoEnciso** [comment]
@alan-j-hu Thanks very much for your valuable input on how to build the Ocaml bindings and for your effort on running the Ocaml tests.
